crates/core/tedge_api/src/entity_store.rs

```diff
@@ -41,6 +46,10 @@ impl AsRef<str> for EntityExternalId {
     }
 }
 
+// XXX: As `EntityExternalId` is used as a part of cloudbound MQTT topic, it
+// can't contain characters invalid in topics, i.e. `+` and `#`. ([MQTT-4.7]).
```

Comment:
> But, I get your confusion and may be the solution is that the `ExternalId` should have been just a trait and `C8yExternalId` and `AzureExternalId` structs could have implemented that type and define their own validation rules in their constructors or `From` impls.

This would be a nice follow-up task.